File: DataEngineering/03_Data_Ingestion/projects/golang/05_ftp_file_downloader/main.go
package main

import (
	"io"
	"log"

	"github.com/pkg/sftp"
	"golang.org/x/crypto/ssh"
)

func downloadSFTP(host, user, pkeyPath, remotePath, localPath string) error {
	log.Printf("Connecting to SFTP server %s as %s...", host, user)

	key, err := os.ReadFile(pkeyPath)
	if err != nil {
		return err
	}

	signer, err := ssh.ParsePrivateKey(key)
	if err != nil {
		return err
	}

	config := &ssh.ClientConfig{
		User: user,
		Auth: []ssh.AuthMethod{
			ssh.PublicKeys(signer),
		},
		HostKeyCallback: ssh.InsecureIgnoreHostKey(), // In prod, use a proper host key callback
	}

	client, err := ssh.Dial("tcp", host, config)
	if err != nil {
		return err
	}
	defer client.Close()

	sftpClient, err := sftp.NewClient(client)
	if err != nil {
		return err
	}
	defer sftpClient.Close()

	log.Printf("Downloading %s to %s...", remotePath, localPath)

	remoteFile, err := sftpClient.Open(remotePath)
	if err != nil {
		return err
	}
	defer remoteFile.Close()

	localFile, err := os.Create(localPath)
	if err != nil {
		return err
	}
	defer localFile.Close()

	bytesCopied, err := io.Copy(localFile, remoteFile)
	if err != nil {
		return err
	}

	log.Printf("Successfully downloaded %d bytes to %s", bytesCopied, localPath)
	return nil
}

func main() {
	log.Println("Starting Data Engineering Project: 05_ftp_file_downloader (SFTP Client)")

	host := os.Getenv("SFTP_HOST")
	if host == "" {
		host = "sftp.example.com:22"
	}
	user := os.Getenv("SFTP_USER")
	if user == "" {
		user = "demo"
	}
	pkeyPath := os.Getenv("SFTP_PKEY_PATH")
	if pkeyPath == "" {
		pkeyPath = "./id_rsa"
	}
	remotePath := os.Getenv("SFTP_REMOTE_PATH")
	if remotePath == "" {
		remotePath = "/incoming/data.csv"
	}
	localPath := os.Getenv("LOCAL_DOWNLOAD_PATH")
	if localPath == "" {
		localPath = "./data.csv"
	}

	if _, err := os.Stat(pkeyPath); os.IsNotExist(err) {
		log.Printf("WARNING: Private key not found at %s. Ensure it exists for successful connection.", pkeyPath)
	}

	if err := downloadSFTP(host, user, pkeyPath, remotePath, localPath); err != nil {
		log.Printf("SFTP Download failed: %v", err)
	}
}

File: DataEngineering/03_Data_Ingestion/projects/python/05_ftp_file_downloader/main.py
import os
import logging
import paramiko
from paramiko.ssh_exception import SSHException

logging.basicConfig(level=logging.INFO, format='%(asctime)s - %(levelname)s - %(message)s')

def download_sftp_file(host: str, port: int, user: str, pkey_path: str, remote_path: str, local_path: str):
    """
    Connects to an SFTP server using an SSH key and downloads a file.
    """
    logging.info(f"Connecting to SFTP server {host}:{port} as {user}...")
    
    try:
        # Setup SSH Client
        ssh = paramiko.SSHClient()
        ssh.set_missing_host_key_policy(paramiko.AutoAddPolicy())
        
        # Load private key
        try:
            private_key = paramiko.RSAKey.from_private_key_file(pkey_path)
        except Exception as e:
            logging.error(f"Failed to load SSH private key: {e}")
            raise
            
        ssh.connect(hostname=host, port=port, username=user, pkey=private_key, timeout=10)
        logging.info("SSH connection established. Opening SFTP session...")
        
        sftp = ssh.open_sftp()
        logging.info(f"Downloading {remote_path} to {local_path}...")
        
        sftp.get(remote_path, local_path)
        logging.info("Download complete.")
        
        sftp.close()
        ssh.close()
    except SSHException as e:
        logging.error(f"SSH/SFTP error occurred: {e}")
        raise
    except Exception as e:
        logging.error(f"An unexpected error occurred: {e}")
        raise

def main():
    logging.info('Starting Data Engineering Project: 05_ftp_file_downloader (Paramiko SFTP)')
    host = os.getenv("SFTP_HOST", "sftp.example.com")
    port = int(os.getenv("SFTP_PORT", 22))
    user = os.getenv("SFTP_USER", "demo")
    pkey_path = os.getenv("SFTP_PKEY_PATH", "/path/to/private/key")
    remote_path = os.getenv("SFTP_REMOTE_PATH", "/incoming/data.csv")
    local_path = os.getenv("LOCAL_DOWNLOAD_PATH", "./data.csv")
    
    try:
        # In a real environment, you must provide a valid key path
        if not os.path.exists(pkey_path):
            logging.warning(f"Private key not found at {pkey_path}. Ensure it exists for actual execution.")
            
        download_sftp_file(host, port, user, pkey_path, remote_path, local_path)
    except Exception as e:
        logging.error("SFTP download pipeline failed.", exc_info=True)

if __name__ == '__main__':
    main()
